plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/writer/GcsTransactionLogSynchronizer.java

```diff
@@ -0,0 +1,139 @@
+        PathCodec pathCodec = googleHadoopFileSystem.getGcsFs().getOptions().getPathCodec();
+        try {
+            createStorageObject(newLogEntryPath.toUri(), entryContents, pathCodec, storage);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public boolean isUnsafe()
+    {
+        return false;
+    }
+
+    /**
+     * This method uses {@link Storage} Cloud Storage API client for creating synchronously a storage object
+     * with the specified path and content.
+     * <p>
+     * This approach avoids dealing with `FSDataOutputStream` exposed by the method
+     * `org.apache.hadoop.fs.FileSystem#create(org.apache.hadoop.fs.Path)` and the intricacies of handling
+     * exceptions which may occur while writing the content to the output stream.
+     */
+    private static StorageObject createStorageObject(URI blobPath, byte[] content, PathCodec pathCodec, Storage storage)
+            throws IOException
+    {
+        CreateFileOptions createFileOptions = new CreateFileOptions(false);
+        CreateObjectOptions createObjectOptions = objectOptionsFromFileOptions(createFileOptions);
+        StorageResourceId storageResourceId = pathCodec.validatePathAndGetId(blobPath, false);
+
+        StorageObject object =
+                new StorageObject()
+                        .setContentEncoding(createObjectOptions.getContentEncoding())
+                        .setMetadata(encodeMetadata(createObjectOptions.getMetadata()))
+                        .setName(storageResourceId.getObjectName());
+
+        InputStream inputStream = new ByteArrayInputStream(content, 0, content.length);
+        Storage.Objects.Insert insert = storage.objects().insert(
+                storageResourceId.getBucketName(),
+                object,
+                new InputStreamContent(createObjectOptions.getContentType(), inputStream));
+        // By setting `ifGenerationMatch` setting to `0`, the creation of the blob will succeed only
+        // if there are no live versions of the object.
+        insert.setIfGenerationMatch(0L);
+        insert.getMediaHttpUploader().setDirectUploadEnabled(true);
+        insert.setName(storageResourceId.getObjectName());
+        return insert.execute();
+    }
+
+    /**
+     * Helper for converting from a Map&lt;String, byte[]&gt; metadata map that may be in a
+     * StorageObject into a Map&lt;String, String&gt; suitable for placement inside a
+     * GoogleCloudStorageItemInfo.
+     */
+    private static Map<String, String> encodeMetadata(Map<String, byte[]> metadata)
+    {
+        return Maps.transformValues(metadata, GcsTransactionLogSynchronizer::encodeMetadataValues);
+    }
+
+    // A function to encode metadata map values
+    private static String encodeMetadataValues(byte[] bytes)
+    {
+        return bytes == null ? Data.NULL_STRING : new String(base64Encoder.encode(bytes), UTF_8);
```

Comment:
Adjusted.


This function was copied (and slightly adjusted) from `com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl#encodeMetadataValues` .


> US_ASCII would be more appropriate.
Why ?

Reply:
> > US_ASCII would be more appropriate.
> 
> Why ?

Because `base64Encoder.encode` doesn't produce UTF-8 -encoded text.
It produces bytes that happen to be letters in ASCII, so it _can_ be decoded as UTF-8. It can be decoded as UTF-8, _because_ it's ASCII.